lib/authentication.js

```diff
@@ -0,0 +1,172 @@
+  ],
+  MAX_LENGTH: [
+    login => login.length > MAX_LOGIN_LENGTH,
+    `The login must be less than ${MAX_LOGIN_LENGTH} characters`
+  ]
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: [
+    (login, password) => login.includes(password),
+    'The login can not contain a password'
+  ],
+  PASSWORD_INCLUDES_LOGIN: [
+    (login, password) => password.includes(login),
+    'The password can not contain a login'
+  ]
+};
+
+class AuthenticationStrength {
+  constructor(
+    hints, // Array, [{ required: Array, optional: Array }, ...]
+    bestPercent // Number, percent of best test group
+  ) {
+    this.hints = hints;
+    this.bestPercent = bestPercent;
+  }
+  get valid() {
+    return !!this.bestPercent;
+  }
+  get strength() {
+    if (this.bestPercent < 25) {
+      return 'Very weak';
+    } else if (this.bestPercent < 40) {
+      return 'Weak';
+    } else if (this.bestPercent < 60) {
+      return 'Good';
+    } else if (this.bestPercent < 80) {
+      return 'Strong';
+    } else {
+      return 'Very strong';
+    }
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  testGroups, // Array, [{ required: [testNames], optional: [testNames] }]
+  ...testArgs // Array, [password]/[login]/[login, password]
+  // Returns: AuthenticationStrength instance
+) => {
+  const test = testsNames => {
+    const testsHints = [];
+    testsNames.forEach(testName => {
+      const [test, hint] = tests[testName];
+      if (test(...testArgs)) testsHints.push(hint);
+    });
+    return testsHints;
+  };
+
+  const hints = [];
+  let bestPercent = 0;
+
+  testGroups.forEach(({ required, optional }) => {
+    const requiredHints = test(required);
+    const optionalHints = test(optional);
+
+    if (!requiredHints.length) {
+      const optTestsCount = optional.length;
+      const optHintsCount = optionalHints.length;
+      const percent = 100 * (optTestsCount - optHintsCount) / optTestsCount;
+      bestPercent = Math.max(bestPercent, percent);
+    }
+
+    hints.push({ required: requiredHints, optional: optionalHints });
+  });
+
+  hints.sort((prevHints, hints) => {
+    if (prevHints.required.length < hints.required.length) {
+      return -1;
+    } else if (prevHints.required.length === hints.required.length) {
+      if (prevHints.optional.length <= hints.optional.length) return -1;
+      else return 1;
+    } else {
+      return 1;
+    }
+  });
+
+  return new AuthenticationStrength(hints, bestPercent);
+};
+
+const checkLogin = (
+  // Function that tests the login
+  login, // String, login for testing
+  testGroups // Array, [{ required: [testNames], optional: [testNames] }]
+  // Returns: AuthenticationStrength instance
+) => (
+  makeTest(loginTests, testGroups, login)
+);
+
+const checkPassword = (
+  // Function that tests the password
+  password, // String, password for testing
```

Comment:
s/password for testing/password to test/ ? (If you agree then same in other places)